{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update the code with proper directory/file names and parameters before running. Do not restrict your analyses to this workflow, explore your microbiota."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP1:  Primers removal/trim-paired end reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#\n",
    "# Trim forward and reverse complementar reads in N bp\n",
    "#\n",
    "\n",
    "TRIM_12=51\n",
    "TRIM_37=1\n",
    "TRIM_DIR=Results/trimmed_seqs/all_seqs\n",
    "SAMPLE_DIR_12=Samples/Seq1-2\n",
    "SAMPLE_DIR_37=Samples/Seq3-7\n",
    "\n",
    "mkdir -p $TRIM_DIR\n",
    "\n",
    "#trim separately samples from seq1-2 and seq3-7\n",
    "#for d in `find $SAMPLE_DIR_12 -name \"*R1_001.fastq.gz\"`; do\n",
    "#\tsampdir=`basename $d _L001_R1_001.fastq.gz`\n",
    "#\tmkdir -p $TRIM_DIR/$sampdir\n",
    "#done\n",
    "\n",
    "for f in Samples/Seq1-2/*; do #not trimmed 3' yet\n",
    "    fastq=`basename $f`\n",
    "    /home/users/vheidrich/seqtk/seqtk trimfq -e $TRIM_12 -b 21 $f | gzip -c > $TRIM_DIR/$fastq\n",
    "done\n",
    "\n",
    "#for d in `find $SAMPLE_DIR_37 -name \"*R1_001.fastq.gz\"`; do\n",
    "#\tsampdir=`basename $d _L001_R1_001.fastq.gz`\n",
    "#\tmkdir -p $TRIM_DIR/$sampdir\n",
    "#done\n",
    "for f in Samples/Seq3-7/*; do\n",
    "    fastq=`basename $f`\n",
    "    /home/users/vheidrich/seqtk/seqtk trimfq -e $TRIM_37 -b 21 $f | gzip -c > $TRIM_DIR/$fastq\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "TRIM_DIR=Results/trimmed_seqs/all_seqs\n",
    "\n",
    "#Add a G to 0length reads\n",
    "\n",
    "for f in $TRIM_DIR/*fastq.gz; do\n",
    "    fastq=`basename $f`\n",
    "    gunzip < $f | sed 's/^[[:blank:]]*$/G/' | gzip -c > temp.gz\n",
    "    mv temp.gz $TRIM_DIR/$fastq\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP2: Import sequences into artifact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "unset LD_LIBRARY_PATH\n",
    "#\n",
    "#In order to use QIIME 2, your input data must be stored in QIIME 2 artifacts\n",
    "#\n",
    "rm -r Results/trimmed_seqs/all_seqs/.ipynb_checkpoints #remove notebook checkpoints folder\n",
    "TRIM_DIR=Results/trimmed_seqs\n",
    "\n",
    "qiime tools import \\\n",
    "  --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "  --input-path $TRIM_DIR/all_seqs \\\n",
    "  --input-format CasavaOneEightSingleLanePerSampleDirFmt \\\n",
    "  --output-path $TRIM_DIR/demux-paired-end.qza\n",
    "\n",
    "qiime demux summarize \\\n",
    "  --i-data $TRIM_DIR/demux-paired-end.qza \\\n",
    "  --o-visualization $TRIM_DIR/demux-paired-end.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP3: DADA2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "unset LD_LIBRARY_PATH\n",
    "\n",
    "#\n",
    "# This method denoises paired-end sequences, dereplicates them, and filters\n",
    "#  chimeras.The reads will also be joined.\n",
    "#\n",
    "#  --i-demultiplexed-seqs ARTIFACT PATH SampleData[PairedEndSequencesWithQuality]\n",
    "#                                  The paired-end demultiplexed sequences to be\n",
    "#                                  denoised.  [required]\n",
    "#  --p-trunc-len-f INTEGER         Position at which forward read sequences\n",
    "#                                  should be truncated due to decrease in\n",
    "#                                  quality. This truncates the 3' end of the of\n",
    "#                                  the input sequences, which will be the bases\n",
    "#                                  that were sequenced in the last cycles.\n",
    "#                                  Reads that are shorter than this value will\n",
    "#                                  be discarded. After this parameter is\n",
    "#                                  applied there must still be at least a 20\n",
    "#                                  nucleotide overlap between the forward and\n",
    "#                                  reverse reads.\n",
    "#  --p-trunc-len-r INTEGER         Position at which reverse read sequences\n",
    "#                                  should be truncated due to decrease in\n",
    "#                                  quality. This truncates the 3' end of the of\n",
    "#                                  the input sequences, which will be the bases\n",
    "#                                  that were sequenced in the last cycles.\n",
    "#                                  Reads that are shorter than this value will\n",
    "#                                  be discarded. After this parameter is\n",
    "#                                  applied there must still be at least a 20\n",
    "#                                  nucleotide overlap between the forward and\n",
    "#                                  reverse reads.\n",
    "#  --o-table ARTIFACT PATH FeatureTable[Frequency]\n",
    "#                                  The resulting feature table. \n",
    "#  --o-representative-sequences ARTIFACT PATH FeatureData[Sequence]\n",
    "#                                  The resulting feature sequences. Each\n",
    "#                                  feature in the feature table will be\n",
    "#                                  represented by exactly one sequence, and\n",
    "#                                  these sequences will be the joined paired-\n",
    "#                                  end sequences.  [required if not passing\n",
    "#                                  --output-dir]\n",
    "#  --o-denoising-stats ARTIFACT PATH SampleData[DADA2Stats]\n",
    "#                                  [required if not passing --output-dir]\n",
    "#\n",
    "\n",
    "\n",
    "DADAED_DIR=Results/trimmed_joined_dadaed_seqs\n",
    "mkdir -p $DADAED_DIR\n",
    "\n",
    "IARTIFACT=Results/trimmed_seqs/demux-paired-end.qza\n",
    "OARTIFACT=$DADAED_DIR/demux-joined-dadaed-fs.qza\n",
    "OARTIFACT1=$DADAED_DIR/demux-joined-dadaed-ft.qza\n",
    "VARTIFACT=$DADAED_DIR/demux-joined-dadaed-stats.qza\n",
    "VARTIFACT1=$DADAED_DIR/demux-joined-dadaed.qzv\n",
    "\n",
    "#p-truncs based on demux-paired-end.qzv\n",
    "qiime dada2 denoise-paired \\\n",
    "  --i-demultiplexed-seqs $IARTIFACT \\\n",
    "  --p-trunc-len-f 0 \\\n",
    "  --p-trunc-len-r 0 \\\n",
    "  --p-max-ee-f 4.0 \\\n",
    "  --p-max-ee-r 4.0 \\\n",
    "  --o-representative-sequences $OARTIFACT \\\n",
    "  --p-n-threads 12 \\\n",
    "  --o-table $OARTIFACT1 \\\n",
    "  --verbose \\\n",
    "  --o-denoising-stats $VARTIFACT\n",
    "\n",
    "#summarize feature table\n",
    "qiime feature-table summarize \\\n",
    "    --i-table $OARTIFACT1 \\\n",
    "    --o-visualization $VARTIFACT1\n",
    "\n",
    "#to view statistics\n",
    "qiime tools export \\\n",
    "    --input-path $VARTIFACT \\\n",
    "    --output-path $DADAED_DIR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP4: Chimeras identification and removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "unset LD_LIBRARY_PATH\n",
    "#\n",
    "# Apply the vsearch uchime_ref method to identify chimeric feature\n",
    "# sequences. The results of this method can be used to filter chimeric\n",
    "# features from the corresponding feature table.\n",
    "#\n",
    "#--i-sequences The feature sequences to be chimera-checked.\n",
    "#--i-table Feature table\n",
    "#--i-reference-sequences The non-chimeric reference sequences. 99% reference used\n",
    "#--o-chimeras The chimeric sequences\n",
    "#--o-nonchimeras The non-chimeric sequences\n",
    "#--o-stats Summary statistics from chimera checking.\n",
    "#\n",
    "\n",
    "CHIMERAS_DIR=Results/trimmed_joined_dadaed_chimeric_seqs\n",
    "DADAED_DIR=Results/trimmed_joined_dadaed_seqs\n",
    "REF_DIR=/home/users/vheidrich/projects_current/Metagenomics/Silva/referenceOTUs_artifacts\n",
    "mkdir -p $CHIMERAS_DIR\n",
    "\n",
    "IARTIFACT=$DADAED_DIR/demux-joined-dadaed-fs.qza\n",
    "IARTIFACT1=$DADAED_DIR/demux-joined-dadaed-ft.qza\n",
    "IARTIFACT2=$REF_DIR/referenceOTUs_99.qza\n",
    "OARTIFACT=$CHIMERAS_DIR/demux-joined-dadaed-chimeras-fs.qza\n",
    "OARTIFACT1=$CHIMERAS_DIR/demux-joined-dadaed-nonchimera-fs.qza\n",
    "VARTIFACT=$CHIMERAS_DIR/demux-joined-dadaed-chimeras_stats.qza\n",
    "VARTIFACT1=$CHIMERAS_DIR/demux-joined-dadaed-chimeras-fs.qzv\n",
    "\n",
    "#99\n",
    "qiime vsearch uchime-ref \\\n",
    "    --i-sequences $IARTIFACT \\\n",
    "    --i-table $IARTIFACT1 \\\n",
    "    --i-reference-sequences $IARTIFACT2 \\\n",
    "    --o-chimeras $OARTIFACT \\\n",
    "    --o-nonchimeras $OARTIFACT1 \\\n",
    "    --p-threads 16 \\\n",
    "    --o-stats $VARTIFACT\n",
    "\n",
    "qiime feature-table tabulate-seqs \\\n",
    "    --i-data $OARTIFACT \\\n",
    "    --o-visualization $VARTIFACT1\n",
    "\n",
    "#qiime tools export \\\n",
    "#    --input-path $IARTIFACT1 \\\n",
    "#    --output-path $DADAED_DIR\n",
    "\n",
    "#biom convert -i $DADAED_DIR/feature-table.biom -o $DADAED_DIR/feature-table.txt --to-tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "unset LD_LIBRARY_PATH\n",
    "#\n",
    "#Filter features from table based on frequency and/or metadata.\n",
    "#\n",
    "#--i-table The feature table from which features should be filtered.\n",
    "#--i-data The sequences from which features should be filtered.\n",
    "#--m-metadata-file Metadata file or artifact viewable as metadata. This option may be supplied multiple \n",
    "#                  times to merge metadata. Feature metadata used with `where` parameter when selecting \n",
    "#                  features to retain, or with `exclude_ids` when selecting features to discard.\n",
    "#--o-filtered-table The resulting filtered table.\n",
    "#--o-filtered-data The resulting filtered sequences.\n",
    "#--o-visualization Visual and tabular summaries of a feature table.\n",
    "#\n",
    "\n",
    "CHIMERIC_DIR=Results/trimmed_joined_dadaed_chimeric_seqs\n",
    "CHIMERAS_DIR=Results/trimmed_joined_dadaed_chimeraremoved_seqs\n",
    "DADAED_DIR=Results/trimmed_joined_dadaed_seqs\n",
    "mkdir -p $CHIMERAS_DIR\n",
    "\n",
    "IARTIFACT=$DADAED_DIR/demux-joined-dadaed-fs.qza\n",
    "IARTIFACT1=$CHIMERIC_DIR/demux-joined-dadaed-nonchimera-fs.qza\n",
    "IARTIFACT2=$DADAED_DIR/demux-joined-dadaed-ft.qza\n",
    "OARTIFACT=$CHIMERAS_DIR/demux-joined-dadaed-chimeraremoved-ft.qza\n",
    "OARTIFACT1=$CHIMERAS_DIR/demux-joined-dadaed-chimeraremoved-fs.qza\n",
    "OARTIFACT2=$CHIMERAS_DIR/demux-joined-dadaed-chimeraremoved-ft.qzv\n",
    "\n",
    "qiime feature-table filter-features \\\n",
    "  --i-table $IARTIFACT2 \\\n",
    "  --m-metadata-file $IARTIFACT1 \\\n",
    "  --o-filtered-table $OARTIFACT\n",
    "  \n",
    "qiime feature-table filter-seqs \\\n",
    "  --i-data $IARTIFACT \\\n",
    "  --m-metadata-file $IARTIFACT1 \\\n",
    "  --o-filtered-data $OARTIFACT1\n",
    "  \n",
    "qiime feature-table summarize \\\n",
    "  --i-table $OARTIFACT \\\n",
    "  --o-visualization $OARTIFACT2\n",
    "\n",
    "#qiime tools export \\\n",
    "#    --input-path $OARTIFACT \\\n",
    "#    --output-path $CHIMERAS_DIR\n",
    "\n",
    "#biom convert -i $CHIMERAS_DIR/feature-table.biom -o $CHIMERAS_DIR/feature-table.txt --to-tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP5: Taxonomic assignment "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "unset LD_LIBRARY_PATH\n",
    "#\n",
    "# explore the taxonomic composition of the samples and relate that to sample metadata.\n",
    "#\n",
    "#--i-classifier Naive Bayes classifier trained on Greengenes 99%\n",
    "#--m-metadata-file mapping file\n",
    "#--o-visualization output visualization path\n",
    "#--i-reads OTUs representative sequences\n",
    "#--i-table OTU table\n",
    "#--o-clasification/--i-taxonomy/--i-input-file artifact with sequences assigned to taxonomy\n",
    "#\n",
    "\n",
    "REF_DIR=/home/users/vheidrich/projects_current/Metagenomics/Silva/referenceOTUs_artifacts\n",
    "TAXA_DIR=Results/taxa_summary\n",
    "#mkdir -p $TAXA_DIR\n",
    "\n",
    "IARTIFACT=Results/trimmed_joined_dadaed_chimeraremoved_seqs/demux-joined-dadaed-chimeraremoved-fs.qza\n",
    "OARTIFACT=$TAXA_DIR/taxonomy_vs.qza\n",
    "VARTIFACT=$TAXA_DIR/taxonomy_vs.qzv\n",
    "REF=$REF_DIR/referenceOTUs_99.qza\n",
    "TAX=$REF_DIR/taxonomy_99.qza\n",
    "\n",
    "#taxonomy assignment\n",
    "qiime feature-classifier classify-consensus-vsearch \\\n",
    "  --i-query $IARTIFACT \\\n",
    "  --i-reference-reads $REF \\\n",
    "  --i-reference-taxonomy $TAX \\\n",
    "  --p-threads 180 \\\n",
    "  --o-classification $OARTIFACT\n",
    "\n",
    "#generate a visualization of the resulting mapping from sequence to taxonomy\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file $OARTIFACT \\\n",
    "  --o-visualization $VARTIFACT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP6: Removal of non-bacterial ASVs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "unset LD_LIBRARY_PATH\n",
    "\n",
    "ASV=Results/trimmed_joined_dadaed_chimeraremoved_seqs/demux-joined-dadaed-chimeraremoved-ft.qza\n",
    "ASV_filtered=Results/trimmed_joined_dadaed_chimeraremoved_seqs/ASV_table_final.qza\n",
    "qiime feature-table filter-features \\\n",
    "  --i-table $ASV \\\n",
    "  --m-metadata-file Results/taxa_summary/taxonomy_vs.qza \\\n",
    "  --p-where \"Taxon='Unassigned' OR Taxon='D_0__Bacteria;D_1__Cyanobacteria;D_2__Oxyphotobacteria;D_3__Chloroplast' OR Taxon='D_0__Bacteria;D_1__Proteobacteria;D_2__Alphaproteobacteria;D_3__Rickettsiales;D_4__Mitochondria'\" \\\n",
    "  --p-exclude-ids \\\n",
    "  --o-filtered-table $ASV_filtered\n",
    "  \n",
    "qiime feature-table summarize \\\n",
    "  --i-table $ASV_filtered \\\n",
    "  --o-visualization Results/trimmed_joined_dadaed_chimeraremoved_seqs/ASV_table_final.qzv\n",
    "\n",
    "ASV=Results/trimmed_joined_dadaed_chimeraremoved_seqs/demux-joined-dadaed-chimeraremoved-fs.qza\n",
    "ASV_filtered=Results/trimmed_joined_dadaed_chimeraremoved_seqs/ASV_seqs_final.qza\n",
    "\n",
    "qiime feature-table filter-seqs \\\n",
    "  --i-data $ASV \\\n",
    "  --m-metadata-file Results/taxa_summary/taxonomy_vs.qza \\\n",
    "  --p-where \"Taxon='Unassigned' OR Taxon='D_0__Bacteria;D_1__Cyanobacteria;D_2__Oxyphotobacteria;D_3__Chloroplast' OR Taxon='D_0__Bacteria;D_1__Proteobacteria;D_2__Alphaproteobacteria;D_3__Rickettsiales;D_4__Mitochondria'\" \\\n",
    "  --p-exclude-ids \\\n",
    "  --o-filtered-data $ASV_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP7: Removal of <3,000 reads samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "unset LD_LIBRARY_PATH\n",
    "\n",
    "ASV=Results/trimmed_joined_dadaed_chimeraremoved_seqs/ASV_table_final.qza\n",
    "ASV_filtered=Results/trimmed_joined_dadaed_chimeraremoved_seqs/ASV_table_final_goodsamples.qza\n",
    "qiime feature-table filter-samples \\\n",
    "  --i-table $ASV \\\n",
    "  --p-min-frequency 3000 \\\n",
    "  --o-filtered-table $ASV_filtered"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
