Here we provide files associated with [our paper](https://doi.org/xxx):

1. Jupyter notebook containing the bioinformatics pipeline used to process sequencing data* (`oral-microbiota-hsct_script.ipynb`)
2. Several R scripts used for plotting and statistics (`R scripts` dir)
3. QIIME 2 artifacts necessary to rerun the analyses (`qza` dir)
4. Metadata and clinical data (`data` dir):
    - Mapping [file](https://github.com/vitorheidrich/oral-microbiota-hsct/blob/main/data/mapping_tmo_0.tsv)
    - Clinical [data]()
    - ATB usage [data]()
    - Blood cell count [data]()
    - [List]() of blooming events
    - ATB usage [variables/classes]()
    - Microbiota [variables/classes]()

*Raw sequencing data have been deposited in the European Nucleotide Archive (ENA) at EMBL-EBI under accession number [PRJEB53914](https://ebi.ac.uk/ena/browser/view/PRJEB53914). Some samples (analyzed in previous studies) were deposited previously in ENA at EMBL-EBI under accession numbers: [PRJEB42862](https://ebi.ac.uk/ena/browser/view/PRJEB42862), [PRJEB49175](https://ebi.ac.uk/ena/browser/view/PRJEB49175). Overall, out of the 444 sequenced samples analyzed in this study, PRJEB53914 contains 315,  PRJEB42862 contains 90, and PRJEB49175 contains 39 samples. Refer to the mapping [file](https://github.com/vitorheidrich/oral-microbiota-hsct/blob/main/data/mapping_tmo_0.tsv) to associate the universal sample IDs with the metadata used in this study.

If you use this data in a publication, please cite as:

Heidrich, V., et al. (2022) XXXXXXX
