Here we provide files associated with [our paper](https://doi.org/xxx):

1. Jupyter notebook containing the bioinformatics pipeline used to process sequencing data* (`oral-microbiota-hsct_script.ipynb`)
2. Several R scripts used for plotting and statistics (`R scripts` dir)
3. QIIME 2 artifacts necessary to rerun the analyses (`qza` dir)
4. Metadata and clinical data (`data` dir)

*Raw sequencing data have been deposited in the European Nucleotide Archive (ENA) at EMBL-EBI under accession number XXXXX (https://ebi.ac.uk/ena/browser/view/XXXXX).

If you use this data in a publication, please cite as:

Heidrich, V., et al. (2022) XXXXXXX
